### PR TITLE
sensor: overflow in sensor_value from double

### LINKED ISSUE #39176: overflow in sensor_value_from_double
**Describe the bug**
The following function will cause a unnecessary overflow of the fractional part (`val2`) when input is more then `2147.493647 ` (`inp >=  INT32_MAX/1000000.0 + 0.01`).
https://github.com/zephyrproject-rtos/zephyr/blob/d198da7bb4bafdd185df5122150a843c1dcea5b1/include/drivers/sensor.h#L670-L675

**To Reproduce**
```C
int main(int argc, char *argv[])
{
    double fp_in = 65536.65536; // UINT16_MAX
    struct sensor_value sv;
    sensor_value_from_double(&sv, fp_in);

    double fp_out = sensor_value_to_double(&sv);
    printf("fp_in=%f, fp_out=%f, val1=%d, val2=%d\n", fp_in, fp_out, sv.val1, sv.val2);
}
```
output:
`fp_in=65536.655360, fp_out=65536.483647, val1=65536, val2=483647`

**Expected behavior**
output from above example should be:
`fp_in=65536.655360, fp_out=65536.655360, val1=65536, val2=655360`

**Impact**
Possible hidden error as converting back to double with `sensor_value_to_double` might not reveal the overflow

**Environment (please complete the following information):**
 - Commit main/master branch

**Suggested fix**
```C
static inline int sensor_value_from_double(struct sensor_value *val, double inp)
{
    if (inp < INT32_MIN || inp > INT32_MAX) {
        return -ERANGE;
    }

    double val2 = (inp - (int32_t) inp) * 1000000;
    if (val2 < INT32_MIN || val2 > INT32_MAX) {
        return -ERANGE;
    }

    val->val1 = (int32_t) inp;
    val->val2 = (int32_t) val2;
    
    return 0;
}
```
I assume adding a return value will not break anything

### PR DESCRIPTION
**Pull request by 5frank**
Fixes #39176

Signed-off-by: Simon Frank <simon.frank@lohmega.com>

### CODE REVIEW

**Review by MaureenHelm — COMMENTED** [review]
Thanks for the fix!

Please review and fix the compliance check failure

**Review by avisconti — COMMENTED** [review]
why not squashing the two commits?

**Review by MaureenHelm — DISMISSED** [review]

**Review by teburd — CHANGES_REQUESTED** [review]
ina219, stm32_temp, and i3g4250 use this call and are not expecting a fallible call. Those drivers should be updated to check for the error and return errors themselves or cast the return to (void) to annotate the call as ignoring the errno if its impossible to fail.

**Review by teburd — DISMISSED** [review]

**Comment by 5frank** [comment]
Not sure why Documentation build was cancelled? I doesn't look like it relates to any of the changes in this PR?

**Comment by gmarull** [comment]
> Not sure why Documentation build was cancelled? I doesn't look like it relates to any of the changes in this PR?

Just rebase, the doc issue was fixed a few weeks ago.

**Review by ycsin — DISMISSED** [review]

**Comment by dleach02** [comment]
@5frank, can you resolve the conflict?

**Comment by henrikbrixandersen** [comment]
@5frank Please rebase on `main`, fix the merge conflicts, and we'll see if that fixes the CI issue as well.

File: include/drivers/sensor.h [hunk 1]
```diff
@@ -666,11 +666,24 @@ static inline double sensor_value_to_double(const struct sensor_value *val)
...
 {
-	val->val1 = (int32_t) inp;
-	val->val2 = (int32_t)(inp * 1000000) % 1000000;
+	if (inp < INT32_MIN || inp > INT32_MAX) {
+		return -ERANGE;
+	}
+
+	double val2 = (inp - (int32_t)inp) * 1000000.0;
+
+	if (val2 < INT32_MIN || val2 > INT32_MAX) {
```
Comment (ycsin) [inline include/drivers/sensor.h:679]: I wonder if the `val2` (same goes for the `inp`) above is treated as an `int` or a `double` by the compiler in the comparison here?

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**